src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,445 @@
+  = ()
+
+let trace_subtract_snoc_left' (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires Snoc? tr2 /\ (let Snoc hd _ = tr2 in tr1 <$ hd))
+    (ensures (
+      let Snoc hd e = tr2 in
+      (tr2 <--> tr1) == Snoc (hd <--> tr1) e
+    ))
+  = ()
+
+let trace_concat_snoc_right (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (ensures tr1 <++> (Snoc tr2 e) == Snoc (tr1 <++> tr2) e)
+  = ()
+
+let snoc_is_concat_singleton (#label_t:Type) (tr :trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (ensures Snoc tr e == (tr <++> (Snoc Nil e)))
+  = ()
+
+let rec trace_concat_grows (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures tr1 <$ (tr2 <++> tr3))
+    [SMTPat (tr1 <$ (tr2 <++> tr3)); SMTPat (tr1 <$ tr2)]
+  = match tr3 with
+    | Nil -> ()
+    | Snoc hd e -> trace_concat_grows tr1 tr2 hd
+
+
+/// Properties connecting trace concatenation, splitting, and subtraction.
+
+let rec trace_split_concat (#label_t:Type) (tr:trace_ label_t) (ts:timestamp{ts < length tr})
+  : Lemma
+    (ensures (
+      let (tr1, e, tr2) = trace_split tr ts in
+      (Snoc tr1 e) <++> tr2 == tr
+    ))
+  = let Snoc hd e = tr in
+    if ts = last_ts tr then ()
+    else trace_split_concat hd ts
+
+let rec trace_subtract_matches_split (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2 /\ Snoc? tr1)
+    (ensures (
+      let (_, _, tl) = trace_split tr2 (length tr1 - 1) in
+      tl == (tr2 <--> tr1)
+    ))
+  = let Snoc hd e = tr2 in
+    let (_, _, tl) = trace_split tr2 (length tr1 - 1) in
+    grows_cases tr1 tr2;
+    eliminate tr1 == tr2 \/ tr1 <$ hd
+    returns tl == (tr2 <--> tr1)
+    with _. ()
+    and _. trace_subtract_matches_split tr1 hd
+
+let rec trace_subtract_concat_right (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires tr2 <$ tr1)
+    (ensures tr2 <++> (tr1 <--> tr2) == tr1)
+    [SMTPat (tr2 <++> (tr1 <--> tr2))]
+  = grows_cases tr2 tr1;
+    eliminate tr1 == tr2 \/ (Snoc? tr1 /\ (let Snoc hd _ = tr1 in tr2 <$ hd))
+    returns tr2 <++> (tr1 <--> tr2) == tr1
+    with _. ()
+    and _. begin
+      let Snoc hd e = tr1 in
+      trace_subtract_concat_right hd tr2;
+      trace_subtract_snoc_left' tr2 tr1;
+      trace_concat_snoc_right tr2 (hd <--> tr2) e
+    end
+    (* Proof with original trace_subtract definition
+    TODO: Remove when trace subtraction is resolved
+    begin
+      trace_split_matches_prefix tr1 (length tr2 - 1);
+      prefix_full_eq tr2;
+      trace_split_concat tr1 (length tr2 - 1);
+      ()
+    end
+*)
+
+let rec trace_subtract_concat_left (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures ((tr2 <--> tr1) <++> tr3) == ((tr2 <++> tr3) <--> tr1))
+  = match tr3 with
+    | Nil -> ()
+    | Snoc hd e -> begin
+      trace_subtract_concat_left tr1 tr2 hd;
+      trace_subtract_snoc_left' tr1 (tr2 <++> tr3)
+    end
```

Comment:
I think this is equivalent to the simpler lemma

```fstar
let rec trace_concat_subtract (#label_t:Type) (tr1 tr2: trace_ label_t)
  : Lemma
    (tr2 == ((tr1 <++> tr2) <--> tr1))
    [SMTPat ((tr1 <++> tr2) <--> tr1)]
  = match tr2 with
    | Nil -> ()
    | Snoc hd e -> begin
      trace_concat_subtract tr1 hd;
      trace_subtract_snoc_left' tr1 (tr1 <++> tr2)
    end
```
(need the improved `concat_grows` lemma)

Then we can prove the current lemma with the simplified one using:
```fstar
let trace_subtract_concat_left (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
  : Lemma
    (requires tr1 <$ tr2)
    (ensures ((tr2 <--> tr1) <++> tr3) == ((tr2 <++> tr3) <--> tr1))
  = 
  //trace_subtract_concat_right tr2 tr1;
  trace_concat_assoc tr1 (tr2 <--> tr1) tr3;
  //trace_subtract_concat_right (tr2 <++> tr3) tr1;
  //trace_concat_subtract tr1 ((tr2 <++> tr3) <--> tr1);
  //trace_concat_subtract tr1 ((tr2 <--> tr1) <++> tr3);
  ()
```
(commented lemmas are triggered by SMT pattern so the proof is one line)